Fix ProjectSettings name casing options and missing String methods

priahoud opened the pull request:
Godot gained the ability to automatically switch to various name-casing schemes in ProjectSettings. The Options in those Settings however are not consistent. While Node Naming is missing a "kebab-case" option, Scene- and Script Naming are missing a "camelCase" Option. 

Also while "to_pascal_case", "to_camel_case" and "to_snake_case" are available as String and StringName Methods, "to_kebab_case" is not.

This patch tries to fix above issues.

- refactored and renamed String::_camelcase_to_underscore to String:_separate_compound_words
- refactored String::to_snake_case to work with the refactored String::_separate_compound_words
- created String::to_kebab_case using the new String::_separate_compound_words
- created corresponding Documentation in String and StringName
- hooked up the ConnectionsDialog
- simplified both switch statements in EditorNode and ProjectDialog
- created and refactored Mono Runtime Interop Functions
- added new kebab-casing Option for Node Names in ProjectSettings
- added missing camelCase Options to Scene- and Node-Names in ProjectSettings
- created additional Unit Tests

Closes #99586

priahoud committed:
Add Options, Functions and Settings to convert Node-Names and Strings to kebab-case

- refactored and renamed String::_camelcase_to_underscore to String:_separate_compound_words
- refactored String::to_snake_case to work with the refactored String::_separate_compound_words
- created char_utils::is_hyphen to catch all hyphen variants in kebab-case conversion
- created String::to_kebab_case using the new String::_separate_compound_words
- created corresponding Documentation in String and StringName
- simplified both switch statements in EditorNode and ProjectDialog
- added new kebab-casing Option for Node Names in ProjectSettings
- added missing camelCase Options to Scene- and Node-Names in ProjectSettings
- simplified Mono RuntimeInterop Functions
- hooked up the ConnectionsDialog
- created additional Unit Tests

priahoud commented:
Note that the Scene- and the Script-Options changed their Enum Indices in order to be consistent with the Node Options, which unfortunately means that existing ProjectSettings change their Meaning.

I opted to be consistent here, but this can of course be changed to be Backward compatible with existing Projects.

AThousandShips commented:
I don't think consistency matters enough here that it justifies changing the order, they don't align anyway as is so we should maintain value ordering IMO

priahoud commented:
> I don't think consistency matters enough here that it justifies changing the order, they don't align anyway as is so we should maintain value ordering IMO

Allright. I will change it to be Backwards compatible then.

priahoud commented:
Rebased and tested - still works as expected

tetrapod00 commented:
See also https://github.com/godotengine/godot-proposals/issues/10727. The existing casing options seem to the ones which are actually used as part of some recommeneded style (plus `kebab-case` which *might* been added speculatively in https://github.com/godotengine/godot/pull/78119; I can't find any issues, PRs, or proposals that specifically want `kebab-case` available).

Unifying the available casing options in the various project settings seems *fine* but be careful you're not adding a casing option to a place where it shouldn't be used for some reason

priahoud commented:
Thanks for the heads up, @tetrapod00!

> See also [godotengine/godot-proposals#10727](https://github.com/godotengine/godot-proposals/issues/10727). The existing casing options seem to the ones which are actually used as part of some recommeneded style

I build this patch out of necessity to comply with our internal naming conventions at work where we use kebab-case everywhere, and since it doesn't really matter how resources, scripts or objects are named in Godot, except for some best practices guidelines (GDScript) or language-specific naming conventions (C#), I opted to include the missing options for all cases - namely kebab-case for nodes and camel-case for node and scenes. The rest is just some re-factoring, integration, documentation and additional tests.

> (plus `kebab-case` which _might_ been added speculatively in #78119;

Correct me if I am wrong, but that PR is already merged and this one just builds on top of it and expands the possible options.

> I can't find any prior issues, PRs, or proposals that specifically want `kebab-case` available).
>
>Unifying the available casing options in the various project settings seems _fine_ but be careful you're not adding a casing option to a place where it shouldn't be used for some reason

See above. We use it internally at work in multiple projects and are happy with it, but I can totally understand if this is too much of a stretch.

tetrapod00 commented:
Seems good then. I don't really have an objection to unifying the casing options available in these settings.

> that PR is already merged

Yeah, and it looked like the `kebab-case` was added speculatively instead of driven by known demand, which is unusual for Godot. But it seemed to have worked out, and if it's being used it should be supported across the multiple settings.

Mickeon commented on doc/classes/String.xml:
```diff
@@ -1010,6 +1010,25 @@
+		<method name="to_kebab_case" qualifiers="const">
+			<return type="String" />
+			<description>
+				Returns the string converted to [code]kebab_case[/code].
```
```suggestion
				Returns the string converted to [code]kebab-case[/code].
```

Mickeon commented on doc/classes/StringName.xml:
```diff
@@ -918,6 +918,25 @@
+		<method name="to_kebab_case" qualifiers="const">
+			<return type="String" />
+			<description>
+				Returns the string converted to [code]kebab_case[/code].
```
```suggestion
				Returns the string converted to [code]kebab-case[/code].
```

Mickeon reviewed (commented):
I can't say whether this is desired or not, but the docs are completely fine and align with the existing `to_snake_case()` description.

RedMser commented on core/string/ustring.cpp:
```diff
@@ -1104,7 +1104,7 @@ const char32_t *String::get_data() const {
 }
 
-String String::_camelcase_to_underscore() const {
+String String::_separate_compound_words() const {
```
Since this method is used in multiple places now, could you consider making it return a PackedStringArray and using the join method on String to concat with underscores/dashes/spaces? Not sure if it would be much of an improvement, but doing multiple replaces on the entire string feels bad.

priahoud replied:
Sounds reasonable. I'm on it!

priahoud replied:
Allright. I think the latest commit should address this suggestion. I also got rid of the multiple string replacements by just looping once over all characters and overwrite them if they match one of the criteria.

Now there is also the question of what to do with the c# side. It has it's own implementation of [SeparateCompoundWords](https://github.com/priahoud/godot/blob/0c0714fd98415efa0c71b7d14e82a63ef697f974/modules/mono/glue/GodotSharp/GodotSharp/Core/StringExtensions.cs#L387), which is only used in [Capitalize](https://github.com/priahoud/godot/blob/0c0714fd98415efa0c71b7d14e82a63ef697f974/modules/mono/glue/GodotSharp/GodotSharp/Core/StringExtensions.cs#L315). [ToPascalCase](https://github.com/priahoud/godot/blob/0c0714fd98415efa0c71b7d14e82a63ef697f974/modules/mono/glue/GodotSharp/GodotSharp/Core/StringExtensions.cs#L353), [ToCamelCase](https://github.com/priahoud/godot/blob/0c0714fd98415efa0c71b7d14e82a63ef697f974/modules/mono/glue/GodotSharp/GodotSharp/Core/StringExtensions.cs#L340), [ToSnakeCase](https://github.com/priahoud/godot/blob/0c0714fd98415efa0c71b7d14e82a63ef697f974/modules/mono/glue/GodotSharp/GodotSharp/Core/StringExtensions.cs#L366) and [ToKebabCase](https://github.com/priahoud/godot/blob/0c0714fd98415efa0c71b7d14e82a63ef697f974/modules/mono/glue/GodotSharp/GodotSharp/Core/StringExtensions.cs#L379) just get their result by marshalling it from the c++ side.

If we do the same with Capitalize, we could just get rid of the SeparateCompoundWords implementation on the  c# side. Overall simplifying things and at the same time making sure both sides behave the same.

RedMser replied:
This new logic is very nice to follow! De-duplicating logic is of course always appreciated, but you'll have to talk to someone who knows more about the C# integration if there are any caveats when doing this, or intentionally different logic between the implementations.

RedMser replied:
> I take it that your concerns are addressed now?

Yep, thanks!

> Ok, I will keep this conversation open and wait for someone from the C# side to comment.

You can also ask on the contributors chat, there is a `#dotnet` channel: https://chat.godotengine.org/

priahoud commented:
Fixed an oversight from the conversion of string::_separate_compound_words where it would still return a pointer to the string itself in case the length of the string was zero. Now it returns an empty PackedStringArray.

I also created char_utils::is_hyphen to catch all hyphen variants in kebab-case conversion and switched the C# implementation of String.Capitalize to just do what ToPascalCase, ToCamelCase, ToSnakeCase and ToKebabCase do - getting the result from the C++ side, which allows us to completely get rid of the custom C# implementation of SeparateCompoundWords. See [this comment](https://github.com/godotengine/godot/pull/94883#discussion_r1855446322) in the discussion above.

KoBeWi commented on scene/main/node.h:
```diff
@@ -108,7 +108,8 @@ class Node : public Object {
 		NAME_CASING_CAMEL_CASE,
-		NAME_CASING_SNAKE_CASE
+		NAME_CASING_SNAKE_CASE,
+		NAME_CASING_KEBAB_CASE
```
```suggestion
		NAME_CASING_KEBAB_CASE,
```

priahoud replied:
Doh! I should have been paying more attention to the surrounding Enums. Thanks!

priahoud commented on core/string/ustring.h:
```diff
@@ -39,6 +39,8 @@
 #include "core/typedefs.h"
 #include "core/variant/array.h"
 
+typedef Vector<String> PackedStringArray;
+
```
I wasn't sure about this one. Thanks for the suggestion!

priahoud commented on core/string/ustring.h:
```diff
@@ -199,7 +201,7 @@ class String {
 	int _count(const String &p_string, int p_from, int p_to, bool p_case_insensitive) const;
 	int _count(const char *p_string, int p_from, int p_to, bool p_case_insensitive) const;
-	String _camelcase_to_underscore() const;
+	PackedStringArray _separate_compound_words() const;
```
Thanks again! New commit incoming.

KoBeWi reviewed (approved):
Functionally it's fine, but the String changes might need to be reviewed by some String expert. Although performance does not matter that much in the context where this is used.

Ivorforce commented on core/string/ustring.cpp:
```diff
@@ -1099,15 +1098,17 @@ const char32_t *String::get_data() const {
-	const char32_t *cstr = get_data();
-	String new_string;
-	int start_index = 0;
+PackedStringArray String::_separate_compound_words() const {
```
I don't think it's a good idea to split the strings upon return. It doesn't improve functionality (as callers could just call `split` if they need it), but it will likely have a very significant performance impact.

priahoud replied:
Hey @Ivorforce , thanks for your feedback!
As it is now, of the five callers of this function, two of them - to_snake_case and to_kebab_case - could get away with just using String::replace on an input string. The other three - capitalize, to_camel_case and to_pascal_case -  would still need to split the input string into an array to get their result - unless there is a more performant way to handle those cases. Maybe building new strings from scratch character by character?
What do you think?

priahoud replied:
I just pushed a new commit where I changed separate_compound_words back to return a string and modeled capitalize, to_camel_case and to_pascal_case after the original code in capitalize. I also rebased everything onto the current master branch. Hope that addresses your concerns.

Ivorforce replied:
Thanks! It definitely looks a lot faster like this :)
There's still a few other places that need a second look, I'll try to comment them all at once in separate discussions.

Ivorforce commented on core/string/ustring.cpp:
```diff
@@ -921,40 +921,72 @@ String String::_camelcase_to_underscore() const {
 
 String String::to_snake_case() const {
-	return _camelcase_to_underscore().replace(" ", "_").strip_edges();
+	return _separate_compound_words().replace(" ", "_");
```
```suggestion
	return _separate_compound_words().replace_char(' ', '_');
```
This was _just_ added so you're going to have to rebase :)

priahoud replied:
Nice! Rebase and changes are done. Build and Tests are running. New commit incoming!

Ivorforce commented on core/string/ustring.cpp:
```diff
@@ -921,40 +921,72 @@ String String::_camelcase_to_underscore() const {
+}
+
+String String::to_kebab_case() const {
+	return _separate_compound_words().replace(" ", "-");
```
```suggestion
	return _separate_compound_words().replace_char(' ', '-');
```

Ivorforce reviewed (approved):
The code looks good to me! Thanks for making the adjustments.
Just needs a rebase, then I think this is ready.

There are several smaller performance concerns still in the code, but it would be a bit unfair to ask you to fix these as they were present in the old code as well. I think what you have is fine :)

Repiteo commented:
Thanks! Congratulations on your first merged contribution! 🎉
